src/ngraph_encapsulate_op.cc

```diff
@@ -165,51 +185,82 @@ class NGraphEncapsulateOp : public tf::OpKernel {
     // Allocate tensors for arguments.
     vector<shared_ptr<ng::runtime::TensorView>> ng_inputs;
 
-    auto& last_used_src_ptrs = m_last_used_src_ptrs_map[ng_function];
-    last_used_src_ptrs.resize(input_shapes.size());
+    std::vector<std::pair<void*, std::shared_ptr<ng::runtime::TensorView>>>&
+        input_caches = m_ng_function_input_cache_map[ng_function];
+    input_caches.resize(input_shapes.size());
 
     for (int i = 0; i < input_shapes.size(); i++) {
       ng::Shape ng_shape(input_shapes[i].dims());
       for (int j = 0; j < input_shapes[i].dims(); ++j) {
         ng_shape[j] = input_shapes[i].dim_size(j);
       }
-
       ng::element::Type ng_element_type;
       OP_REQUIRES_OK(ctx, TFDataTypeToNGraphElementType(ctx->input(i).dtype(),
                                                         &ng_element_type));
 
-      void* src_ptr = (void*)tf::DMAHelper::base(&ctx->input(i));
-      auto t = m_ng_backend->create_tensor(ng_element_type, ng_shape, src_ptr);
-
-      // Mark each tensor as non-stale if:
-      //
-      //   1. the freshness tracker says the tensor has not changed since
-      //      the last time ng_function was called, and
-      //   2. we are using the same tensor in this argument position as
-      //      the one we used last time ng_function was called.
-      if (m_freshness_tracker->IsFresh(src_ptr, ng_function) &&
-          src_ptr == last_used_src_ptrs[i]) {
-        t->set_stale(false);
+      // At the first call of the ng_function, both last_src_ptr and
+      // last_tv shall point to null. Otherwise, they are retrived
+      // from cache.
+      void* last_src_ptr = input_caches[i].first;
```

Comment:
You're right, we only need `last_src_ptr` for `CPU`. However, we'll need to cache`ng::runtime::TensorView` for all backends, and currently they are stored in a pair. So it might be shorter to keep the current way. Otherwise, adding
```cpp
if (m_ng_backend_name == "CPU") {
  void* last_src_ptr = input_caches[i].first;
}
```
here doesn't seem to save us much.

Reply:
Make sense.